pkg/operator/ceph/cluster/controller.go

```diff
@@ -532,6 +536,16 @@ func (c *ClusterController) onUpdate(oldObj, newObj interface{}) {
 
 	cluster.Spec = &newClust.Spec
 
+	// If the image version changed let's make sure we can safely upgrade
+	// check ceph's status, if not healthy we fail
+	if versionChanged {
```

Comment:
If we're only checking for `versionChanged` in the `onUpdate()` method, we may be missing a case for the upgrade. When the operator is restarted, `onAdd()` will be called with the latest state of the CR. So it looks like the upgrade will proceed if the operator is restarted. The only way to catch that case would be for the orchestration to store the version that was orchestrated (eg. in a configmag) so it can be used in a later orchestration. Or else perhaps we could query the pod specs and check if the container image is simply changed from the image in the CR.

Reply:
Are you saying it's a bad thing to trigger the upgrade if the operator is restarted? The problem is that this is how the current logic is built, only the mon deployments won't be re-created but all the other daemons will see their Deployment to be updated. So I'd say, upgrade or not we still need to perform our checks.

Reply:
Let's say the cluster is in HEALTH_ERR state and the user updates the CR with a new ceph version. The update to the CR will not be processed since the cluster is not healthy. Are you saying that if the operator is restarted, the upgrade will proceed anyway? What I'm saying is that this behavior should be consistent. If the user only has to restart the operator to continue with the upgrade in the error health condition, seems like there is no use for the health_err check here.

Reply:
Ack!